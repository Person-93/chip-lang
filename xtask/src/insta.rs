use crate::{CommandExt, ErrorExt, Parser, ResultExt};
use std::{env, process::Command};

/// Run `cargo nextest run` then `cargo insta review`
#[derive(Default, Parser)]
pub struct Cli {}

pub fn main(Cli {}: Cli) -> ! {
  cargo()
    .args(["nextest", "run"])
    .env("INSTA_FORCE_PASS", "1")
    .wait()
    .exit_on_err();
  cargo().args(["insta", "review"]).exec().exit();
}

fn cargo() -> Command {
  env::var_os("CARGO").map_or_else(|| Command::new("cargo"), Command::new)
}
