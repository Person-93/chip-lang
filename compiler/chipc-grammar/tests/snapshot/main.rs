use crate::debug_print::print;
use insta::{assert_snapshot, glob};
use std::fs;
use tree_sitter::Parser;

fn main() {
  let mut parser = Parser::new();
  parser
    .set_language(chipc_grammar::language())
    .expect("assign language to parser");

  glob!("cases/*.chip", |path| {
    parser.reset();
    let source = fs::read_to_string(path).expect("read snapshot test case");
    let tree = parser
      .parse(&source, None)
      .expect("parse snapshot test case");
    assert_snapshot!(print(&tree, &source));
  });
}

mod debug_print {
  use std::fmt::{self, Write};
  use tree_sitter::{Point, Tree, TreeCursor};

  pub fn print(tree: &Tree, source: &str) -> String {
    let mut output = String::new();
    let mut cursor = tree.walk();
    output
      .print_node(&mut cursor, source, 0)
      .expect("debug format tree");
    output
  }

  impl<T: Write> NodePrinter for T {}

  trait NodePrinter: Write {
    fn print_node(
      &mut self,
      cursor: &mut TreeCursor,
      source: &str,
      indentation: u8,
    ) -> fmt::Result {
      let node = cursor.node();
      let kind = node.kind();
      let Point { row, column } = node.start_position();
      let Point {
        row: end_row,
        column: end_column,
      } = node.end_position();

      self.indent(indentation)?;
      write!(self, "{kind} @ {row}:{column}..{end_row}:{end_column}")?;
      if node.is_named() && node.child_count() == 0 {
        write!(self, "\t{}", node.utf8_text(source.as_bytes()).unwrap())?;
      }
      writeln!(self)?;

      if cursor.goto_first_child() {
        self.print_node(&mut cursor.clone(), source, indentation + 1)?;
        while cursor.goto_next_sibling() {
          self.print_node(&mut cursor.clone(), source, indentation + 1)?;
        }
      }

      Ok(())
    }

    fn indent(&mut self, indentation: u8) -> fmt::Result {
      for _ in 0..indentation {
        write!(self, "  ")?;
      }
      Ok(())
    }
  }
}
