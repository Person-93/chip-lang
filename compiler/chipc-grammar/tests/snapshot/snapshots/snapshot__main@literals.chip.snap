---
source: compiler/chipc-grammar/tests/snapshot/main.rs
assertion_line: 18
expression: "print(&tree, &source)"
input_file: compiler/chipc-grammar/tests/snapshot/cases/literals.chip
---
file @ 0:0..26:0
  function @ 0:0..25:1
    fn @ 0:0..0:2
    identifier @ 0:3..0:7	main
    ( @ 0:7..0:8
    ) @ 0:8..0:9
    : @ 0:9..0:10
    codeblock @ 0:11..25:1
      { @ 0:11..0:12
      binding @ 1:2..1:18
        let @ 1:2..1:5
        identifier @ 1:6..1:12	number
        := @ 1:13..1:15
        numeric_literal @ 1:16..1:18
          digits @ 1:16..1:18	42
      ; @ 1:18..1:19
      binding @ 2:2..2:20
        let @ 2:2..2:5
        identifier @ 2:6..2:12	number
        := @ 2:13..2:15
        numeric_literal @ 2:16..2:20
          digits @ 2:16..2:18	42
          u8 @ 2:18..2:20
      ; @ 2:20..2:21
      binding @ 3:2..3:21
        let @ 3:2..3:5
        identifier @ 3:6..3:12	number
        := @ 3:13..3:15
        numeric_literal @ 3:16..3:21
          digits @ 3:16..3:18	42
          u16 @ 3:18..3:21
      ; @ 3:21..3:22
      binding @ 4:2..4:21
        let @ 4:2..4:5
        identifier @ 4:6..4:12	number
        := @ 4:13..4:15
        numeric_literal @ 4:16..4:21
          digits @ 4:16..4:18	42
          u32 @ 4:18..4:21
      ; @ 4:21..4:22
      binding @ 5:2..5:21
        let @ 5:2..5:5
        identifier @ 5:6..5:12	number
        := @ 5:13..5:15
        numeric_literal @ 5:16..5:21
          digits @ 5:16..5:18	42
          u64 @ 5:18..5:21
      ; @ 5:21..5:22
      binding @ 6:2..6:22
        let @ 6:2..6:5
        identifier @ 6:6..6:12	number
        := @ 6:13..6:15
        numeric_literal @ 6:16..6:22
          digits @ 6:16..6:18	42
          u128 @ 6:18..6:22
      ; @ 6:22..6:23
      binding @ 7:2..7:23
        let @ 7:2..7:5
        identifier @ 7:6..7:12	number
        := @ 7:13..7:15
        numeric_literal @ 7:16..7:23
          digits @ 7:16..7:18	42
          usize @ 7:18..7:23
      ; @ 7:23..7:24
      binding @ 8:2..8:20
        let @ 8:2..8:5
        identifier @ 8:6..8:12	number
        := @ 8:13..8:15
        numeric_literal @ 8:16..8:20
          digits @ 8:16..8:18	42
          i8 @ 8:18..8:20
      ; @ 8:20..8:21
      binding @ 9:2..9:21
        let @ 9:2..9:5
        identifier @ 9:6..9:12	number
        := @ 9:13..9:15
        numeric_literal @ 9:16..9:21
          digits @ 9:16..9:18	42
          i16 @ 9:18..9:21
      ; @ 9:21..9:22
      binding @ 10:2..10:21
        let @ 10:2..10:5
        identifier @ 10:6..10:12	number
        := @ 10:13..10:15
        numeric_literal @ 10:16..10:21
          digits @ 10:16..10:18	42
          i32 @ 10:18..10:21
      ; @ 10:21..10:22
      binding @ 11:2..11:21
        let @ 11:2..11:5
        identifier @ 11:6..11:12	number
        := @ 11:13..11:15
        numeric_literal @ 11:16..11:21
          digits @ 11:16..11:18	42
          i64 @ 11:18..11:21
      ; @ 11:21..11:22
      binding @ 12:2..12:22
        let @ 12:2..12:5
        identifier @ 12:6..12:12	number
        := @ 12:13..12:15
        numeric_literal @ 12:16..12:22
          digits @ 12:16..12:18	42
          i128 @ 12:18..12:22
      ; @ 12:22..12:23
      binding @ 13:2..13:23
        let @ 13:2..13:5
        identifier @ 13:6..13:12	number
        := @ 13:13..13:15
        numeric_literal @ 13:16..13:23
          digits @ 13:16..13:18	42
          isize @ 13:18..13:23
      ; @ 13:23..13:24
      binding @ 15:2..15:21
        let @ 15:2..15:5
        identifier @ 15:6..15:12	number
        := @ 15:13..15:15
        numeric_literal @ 15:16..15:21
          digits @ 15:16..15:18	42
          f32 @ 15:18..15:21
      ; @ 15:21..15:22
      binding @ 16:2..16:21
        let @ 16:2..16:5
        identifier @ 16:6..16:12	number
        := @ 16:13..16:15
        numeric_literal @ 16:16..16:21
          digits @ 16:16..16:18	42
          f64 @ 16:18..16:21
      ; @ 16:21..16:22
      binding @ 17:2..17:20
        let @ 17:2..17:5
        identifier @ 17:6..17:12	number
        := @ 17:13..17:15
        numeric_literal @ 17:16..17:20
          digits @ 17:16..17:18	42
          . @ 17:18..17:19
          digits @ 17:19..17:20	0
      ; @ 17:20..17:21
      binding @ 18:2..18:23
        let @ 18:2..18:5
        identifier @ 18:6..18:12	number
        := @ 18:13..18:15
        numeric_literal @ 18:16..18:23
          digits @ 18:16..18:18	42
          . @ 18:18..18:19
          digits @ 18:19..18:20	0
          f32 @ 18:20..18:23
      ; @ 18:23..18:24
      binding @ 19:2..19:23
        let @ 19:2..19:5
        identifier @ 19:6..19:12	number
        := @ 19:13..19:15
        numeric_literal @ 19:16..19:23
          digits @ 19:16..19:18	42
          . @ 19:18..19:19
          digits @ 19:19..19:20	0
          f64 @ 19:20..19:23
      ; @ 19:23..19:24
      binding @ 21:2..21:27
        let @ 21:2..21:5
        identifier @ 21:6..21:10	text
        := @ 21:11..21:13
        string_literal @ 21:14..21:27	"hello world"
      ; @ 21:27..21:28
      binding @ 23:2..23:21
        let @ 23:2..23:5
        identifier @ 23:6..23:13	boolean
        := @ 23:14..23:16
        boolean_literal @ 23:17..23:21
          true @ 23:17..23:21
      ; @ 23:21..23:22
      binding @ 24:2..24:22
        let @ 24:2..24:5
        identifier @ 24:6..24:13	boolean
        := @ 24:14..24:16
        boolean_literal @ 24:17..24:22
          false @ 24:17..24:22
      ; @ 24:22..24:23
      } @ 25:0..25:1
